internal/kafka/internal/workers/cluster_mgrs/dynamic_scale_down_mgr.go

```diff
@@ -0,0 +1,324 @@
+			clusterService:                         m.clusterService,
+			dryRun:                                 !m.dataplaneClusterConfig.EnableDynamicScaleDownManagerScaleDownTrigger,
+			clusterID:                              clusterID,
+			indexesOfStreamingUnitForSameClusterID: existing.indexesOfStreamingUnitForSameClusterID,
+		}
+
+		glog.Infof("evaluating dynamic scale down for cluster '%s'", clusterID)
+		shouldScaleDown, err := dynamicScaleDownProcessor.ShouldScaleDown()
+		if err != nil {
+			errList.AddErrors(err)
+			continue
+		}
+		if shouldScaleDown {
+			glog.Infof("data plane scale down need detected for cluster '%s'", clusterID)
+			err := dynamicScaleDownProcessor.ScaleDown()
+			if err != nil {
+				errList.AddErrors(err)
+				continue
+			}
+		}
+	}
+
+	if errList.IsEmpty() {
+		return nil
+	}
+
+	return errList
+}
+
+func (m *DynamicScaleDownManager) logReconcileEventEnd() {
+	glog.Infoln("dynamic scale down reconcile event finished")
+}
+
+// dynamicScaleDownExecutor is able to perform dynamic ScaleDown execution actions
+type dynamicScaleDownExecutor interface {
+	ScaleDown() error
+}
+
+// dynamicScaleDownEvaluator is able to perform dynamic ScaleDown evaluation actions
+type dynamicScaleDownEvaluator interface {
+	ShouldScaleDown() (bool, error)
+}
+
+// dynamicScaleDownProcessor is able to process dynamic ScaleDown reconcile events
+type dynamicScaleDownProcessor interface {
+	dynamicScaleDownExecutor
+	dynamicScaleDownEvaluator
+}
+
+// standardDynamicScaleDownProcessor is the default dynamicScaleDownProcessor
+// used when dynamic scaling is enabled.
+// It assumes the provided kafkaStreamingUnitCountPerClusterList does not
+// contain any element with a Status attribute with value 'failed'
+type standardDynamicScaleDownProcessor struct {
+	clusterID                    string
+	regionsSupportedInstanceType config.InstanceTypeMap
+	// kafkaStreamingUnitCountPerClusterList must not contain any element
+	// with a Status attribute with value 'failed'
+	kafkaStreamingUnitCountPerClusterList services.KafkaStreamingUnitCountPerClusterList
+	// indexesOfStreamingUnitForSameClusterID is the index of the straming unit in the "kafkaStreamingUnitCountPerClusterList" that has the
+	// same "clusterID" as the one that's current being processed
+	indexesOfStreamingUnitForSameClusterID []int
+	supportedKafkaInstanceTypesConfig      *config.SupportedKafkaInstanceTypesConfig
+	clusterService                         services.ClusterService
+
+	// dryRun controls whether the ScaleDown method performs real actions.
+	// Useful when you don't want to trigger a real scale down.
+	dryRun bool
+}
+
+var _ dynamicScaleDownEvaluator = &standardDynamicScaleDownProcessor{}
+var _ dynamicScaleDownExecutor = &standardDynamicScaleDownProcessor{}
+var _ dynamicScaleDownProcessor = &standardDynamicScaleDownProcessor{}
+
+// ShouldScaleDown indicates whether a new data plane
+// cluster should be created for a given instance type in the given provider
```

Comment:
This is incorrect, I guess due to partial copy of the ShouldScaleUp doc. It is talking about creating a new cluster instead of deleting it, and at a different scope level too

Reply:
Thanks for this. 
This was a heavy copy and paste #ashamed